lxd/daemon.go

```diff
@@ -359,6 +361,220 @@ func allowProjectResourceList(d *Daemon, r *http.Request) response.Response {
+	processEntity := func(objEntityURL, relation string) error {
+		dataIdx, ok := urls[objEntityURL]
+		if !ok {
+			// URL not found, skip
+			return nil
+		}
+
+		switch {
+		case t.Kind() == reflect.Ptr && t.Elem().Kind() == reflect.Struct:
+			// Pointer to struct
+			return appendEntitlement(data, relation)
+		case t.Kind() == reflect.Slice && t.Elem().Kind() == reflect.Ptr:
+			// Slice of pointers to struct
+			if dataIdx >= v.Len() {
+				return fmt.Errorf("Data index out of range while adding entitlements: %v", dataIdx)
+			}
+
+			return appendEntitlement(v.Index(dataIdx).Interface(), relation)
+		case t.Kind() == reflect.Ptr && t.Elem().Kind() == reflect.Slice:
+			// Pointer to slice of struct
+			sliceValue := v.Elem()
+			if dataIdx >= sliceValue.Len() {
+				return fmt.Errorf("Data index out of range while adding entitlements: %v", dataIdx)
+			}
+
+			structPtr := sliceValue.Index(dataIdx).Addr()
+			return appendEntitlement(structPtr.Interface(), relation)
+		}
+
+		return nil
+	}
+
+	if !(t.Kind() == reflect.Ptr && t.Elem().Kind() == reflect.Struct || t.Kind() == reflect.Slice && t.Elem().Kind() == reflect.Ptr || t.Kind() == reflect.Ptr && t.Elem().Kind() == reflect.Slice) {
+		return fmt.Errorf("Data must be either a pointer to struct, slice of pointers to struct, or pointer to slice of struct while adding entitlements, got %v", t.Kind())
+	}
+
+	if t.Kind() == reflect.Ptr && t.Elem().Kind() == reflect.Slice {
+		if t.Elem().Elem().Kind() != reflect.Struct {
+			return fmt.Errorf("Adding entitlements for pointer to slice, the slice elements must be structs, got %v", t.Elem().Elem().Kind())
+		}
+	}
+
+	for k, tuples := range openFGACache.ReadStartingFromUserCache {
+		if tuples == nil {
+			continue
+		}
+
+		var key storage.ReadStartingWithUserFilter
+		err := json.Unmarshal([]byte(k), &key)
+		if err != nil {
+			return fmt.Errorf("Could not unmarshal OpenFGA cache key into 'ReadStartingWithUserFilter': %w", err)
+		}
+
+		if key.ObjectType != entityType.String() {
+			continue
+		}
+
+		for _, tuple := range tuples {
+			objSlice := strings.Split(tuple.Key.Object, ":")
+			if len(objSlice) != 2 {
+				return fmt.Errorf("Could not split OpenFGA tuple key object into entity type and entity URL: %v", tuple.Key.Object)
+			}
+
+			if objSlice[0] != entityType.String() {
+				continue
+			}
+
+			err = processEntity(objSlice[1], tuple.Key.Relation)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, tuples := range openFGACache.ReadUsersetTuplesCache {
+		if tuples == nil {
+			continue
+		}
+
+		for _, tuple := range tuples {
+			objSlice := strings.Split(tuple.Key.Object, ":")
+			if len(objSlice) != 2 {
+				return fmt.Errorf("Could not split OpenFGA tuple key object into entity type and entity URL: %v", tuple.Key.Object)
+			}
+
+			if objSlice[0] != entityType.String() {
+				continue
+			}
+
+			err = processEntity(objSlice[1], tuple.Key.Relation)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
```

Comment:
I'm very confused by this. There are a few things to address here.
1. We can't bypass calls to OpenFGA by accessing the cache directly. We don't know that the cache contains all the necessary information. We will need to call the OpenFGA server itself to check the entitlements on a per-entity basis. The cache should be read from/written to only from in OpenFGA datastore implementation itself.
2. This is leaking a lot of OpenFGA specific logic from the `lxd/db/openfga` and `lxd/auth/drivers` packages. We should try our utmost to keep that logic where it is.
3. Using reflection should be a last resort IMO. If we absolutely require it then we need to have an awful lot of comments to explain why and what it's doing.

For 1, we still need to call the OpenFGA server. I suggest calling `CheckPermission` for each entitlement in the `auth.EntitlementsByEntityType` map value for the entity type. This will be inefficient when listing entitlements for a list of resources (e.g. `GET /1.0/instances?recursion=1&with-entitlements=true`) so I suggest in this calling `auth.GetPermissionChecker` with each entitlement in the `auth.EntitlementsByEntityType` map.

Alternatively, I suggest adding two methods to the `auth.Authorizer` interface to perform the enrichment. One for a single entity which will call `Check` on the OpenFGA server, and one for a list of entities that will call `ListObjects`. The implementation for the TLS authorizer can return a 501 not implemented. The authorizer itself can return a bad request error if the caller is not fine-grained.

For 3, I suggest creating an interface with method `AddEntitlements([]auth.Entitlement)` and implement it for the API types in question (all API types represented by the OpenFGA model).

Reply:
Thanks. I agree with you on the three points. I'll come up with a new approach.

Reply:
@markylaing I updated the approach. What do you think now?

Reply:
The approach looks good :) It might be possible to make the new Authorizer functions more efficient by calling `ListObjects` and `Check` on the OpenFGA server directly. We can revisit that at a later date though, for now re-using `CheckPermission` and `GetPermissionChecker` is reasonable.